x/amm/keeper/update_pool_for_swap.go

```diff
@@ -9,127 +9,152 @@ import (
 	if err != nil {
-		return sdkmath.ZeroInt(), err
+		return err
+	}
+	err = k.AddToPoolBalanceAndUpdateLiquidity(ctx, &pool, sdkmath.ZeroInt(), tokensIn)
+	if err != nil {
+		return err
+	}
+
+	// send tokensOut from pool to sender
+	err = k.bankKeeper.SendCoins(ctx, poolAddr, recipient, tokensOut)
+	if err != nil {
+		return err
+	}
+
+	err = k.RemoveFromPoolBalanceAndUpdateLiquidity(ctx, &pool, sdkmath.ZeroInt(), tokensOut)
+	if err != nil {
+		return err
 	}
 
 	// apply swap fee when weight balance bonus is not available
 	swapFeeInCoins := sdk.Coins{}
-	if !weightBalanceBonus.IsPositive() {
-		swapFeeInCoins = PortionCoins(tokensIn, swapFeeIn)
+	if swapFee.IsPositive() {
+		if givenOut {
+			takeFeesFrom := sdk.NewCoins(sdk.NewCoin(tokenIn.Denom, oracleInAmount))
+			if !pool.PoolParams.UseOracle {
+				takeFeesFrom = tokensIn
+			}
+			// if swapInGivenOut, use oracleIn amount to get swap fees
+			swapFeeInCoins = PortionCoins(takeFeesFrom, swapFee)
+		} else {
+			swapFeeInCoins = PortionCoins(tokensIn, swapFee)
+		}
 	}
 
 	// send swap fee to rebalance treasury
 	if swapFeeInCoins.IsAllPositive() {
 		rebalanceTreasury := sdk.MustAccAddressFromBech32(pool.GetRebalanceTreasury())
 		err = k.bankKeeper.SendCoins(ctx, poolAddr, rebalanceTreasury, swapFeeInCoins)
 		if err != nil {
-			return sdkmath.ZeroInt(), err
+			return err
 		}
+
+		err = k.RemoveFromPoolBalanceAndUpdateLiquidity(ctx, &pool, sdkmath.ZeroInt(), swapFeeInCoins)
+		if err != nil {
+			return err
+		}
+
 		err = k.OnCollectFee(ctx, pool, swapFeeInCoins)
 		if err != nil {
-			return sdkmath.ZeroInt(), err
+			return err
 		}
 	}
 
-	// Send coins to recipient
-	err = k.bankKeeper.SendCoins(ctx, poolAddr, recipient, sdk.Coins{tokenOut})
-	if err != nil {
-		return sdkmath.ZeroInt(), err
-	}
+	// init weightRecoveryFeeAmount to zero
+	weightRecoveryFeeAmount := sdkmath.ZeroInt()
 
-	// apply swap fee when weight balance bonus is not available
-	swapFeeOutCoins := sdk.Coins{}
-	if !weightBalanceBonus.IsPositive() {
-		swapFeeOutCoins = PortionCoins(tokensOut, swapFeeOut)
-	}
-	if swapFeeOutCoins.IsAllPositive() {
+	// send half (weight breaking fee portion) of weight breaking fee to rebalance treasury
+	if pool.PoolParams.UseOracle && weightBalanceBonus.IsNegative() {
+		params := k.GetParams(ctx)
 		rebalanceTreasury := sdk.MustAccAddressFromBech32(pool.GetRebalanceTreasury())
-		err = k.bankKeeper.SendCoins(ctx, poolAddr, rebalanceTreasury, swapFeeOutCoins)
-		if err != nil {
-			return sdkmath.ZeroInt(), err
-		}
-		err = k.OnCollectFee(ctx, pool, swapFeeOutCoins)
-		if err != nil {
-			return sdkmath.ZeroInt(), err
+		// we are multiplying here by params.WeightBreakingFeePortion as we didn't multiply in pool.SwapIn/OutGiveOut/In for weight breaking fee
+		weightRecoveryFee := weightBalanceBonus.Abs().Mul(params.WeightBreakingFeePortion)
+
+		if givenOut {
+			weightRecoveryFeeAmount = oracleInAmount.ToLegacyDec().Mul(weightRecoveryFee).RoundInt()
```

Comment:
I think it should also consider `if !pool.UseOracle` for using tokenIn to oracleInAmount